Trigger buffer events during ads

karimMourra opened the pull request:
### This PR will...
add a listener for `buffer` on instream and trigger a 'buffer' event when an ad is loaded.
### Why is this Pull Request needed?
Currently when a preRoll is loaded, the player state goes from `idle` to `buffering` but a buffer event is never triggered. I think a buffer event should fire because a developer should be able to rely on our events to detect state changes.

This bug was noticed while working on performance improvements for the iOS SDK. In the past, we would use the `getState` API to determine the state. This resulted in a large amount of javascript injections which are costly. To reduce the amount of javascript injections, and since WKWebView handles javascript asynchronously, we have decided to rely on the player's events to track the state. Given that the buffer event doesn't fire on ads, our state tracking gets out of sync. 

If firing buffer events during ad playback damages our event orders, we could look into changing the state in the SDK to buffer on other events such as playAttempt or adStart. Our preference though would be to rely on the player's buffer event. 
### Are there any points in the code the reviewer needs to double check?
Should we have a adBuffer event ?
### Are there any Pull Requests open in other repos which need to be merged with this?
n/a
#### Addresses Issue(s):
SDK-4152

johnBartos commented:
<!--
  0 failure: 
  2 warning:  🛠 There are modi..., 🗿 Set a mileston...
  
  
  DangerID: danger-id-default;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  🛠 There are modified src files, but no test changes. Add tests if you're able to.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  🗿 Set a milestone. It should be the ticket's fix version in JIRA.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

karimMourra commented:
@reviewer: an integration test failed due to flakiness

karimMourra commented:
@pajong `off` is called in the instream's `destroy`. @robwalch thanks, I will address shortly

robwalch commented on src/js/controller/instream-adapter.js:
```diff
@@ -89,6 +89,7 @@ const InstreamAdapter = function(_controller, _model, _view, _mediaPool) {
         _adProgram.setup();
 
         _adProgram.on('all', _instreamForward, this);
+        _adProgram.on('buffer', triggerAdBuffer, this);
```
We already handle state changes in `_handleStateChange`.

pajong replied:
@robwalch Isn't this different from state event? (`buffering` vs `buffer`).
This `buffer` event comes from `program-listener` every time the buffer changes: https://github.com/jwplayer/jwplayer/blob/master/src/js/program/program-listeners.js#L73

robwalch replied:
The point is to match the player state changes.

Player events | Ad mode events
------------ | -------------
buffer | adBuffer
play | adPlay
pause | adPause

karimMourra replied:
i'm confused by `handleStateChange`; the `newState` is coming from the provider, but the `oldState` is coming from the model. But the model's state is set to `buffering` before the provider state change , so `handleStateChange` detects a change from `buffering` to `loading` which is incorrect. Meanwhile the player's state is still `idle`. What should the flow be ?
https://github.com/jwplayer/jwplayer/pull/3169/files#diff-4b5d9c794240aa20cb620353eb027d56R114

robwalch commented on src/js/controller/instream-adapter.js:
```diff
@@ -417,6 +418,10 @@ const InstreamAdapter = function(_controller, _model, _view, _mediaPool) {
     }
 
+    function triggerAdBuffer(evt) {
+        _controller.trigger(AD_BUFFER, evt);
```
Can we deal with this in `_handleStateChange`?

robwalch replied:
Also see `_stateHandler` in ad-program-controller

```
    _stateHandler(state) {
        const { model } = this;
        switch (state) {
            case STATE_PLAYING:
            case STATE_PAUSED:
                model.set(PLAYER_STATE, state);
                break;
            default:
                break;
        }
    }
```

robwalch replied:
and 

```
function normalizeState(newstate) {
    if (newstate === STATE_LOADING || newstate === STATE_STALLED) {
        return STATE_BUFFERING;
    }
    return newstate;
}
```

robwalch replied:
What I expect to see is `_triggerAdPlayPause` renamed to `_triggerAdState` and add handling for buffering states
    function _triggerAdPlayPause(event) {
        const { newstate } = event;
        switch (newstate) {
            STATE_PLAYING:
            _controller.trigger(AD_PLAY, event);
           break;
          STATE_PAUSED:
            _controller.trigger(AD_PAUSE, event);
            break;
          STATE_LOADING:
          STATE_STALLED:
          STATE_BUFFERING:
            _controller.trigger(AD_BUFFER, event); // make any changes necessary to event properties
            break;
        }
    }

karimMourra commented:
@robwalch I addressed your feedback. 
adBuffer does not fire on IMA and Freewheel, I need to set the buffering state on the ad plugins' `instreamProvider` before the adImpression is fired, possibly on the plugin's `loaded` event
